echidna/limit/test_statistic.py

```diff
@@ -0,0 +1,269 @@
+        """
+        return self._type
+
+    def get_spectra_par(self):
+        """
+        Returns:
+          (:class:`echidna.core.spectra.SpectraParameter):The parameter
+           class for the projection dimension, stored in
+           :attr:`_spectra_par`.
+        """
+        return self._spectra_par
+
+    def get_per_bin(self):
+        """
+        Returns:
+          (:class:`numpy.ndarray`): 1D numpy array to hold value
+            of test statistic calculated for each bin from the most
+            recent call to :meth:`compute_statistic`.
+        """
+        return self._per_bin
+
+    # Method adapted from http://codereview.stackexchange.com/a/47115
+    @abc.abstractmethod
+    def compute_statistic(self, observed, expected):
+        """ Compute the value of the test statistic.
+
+        .. warning:: This method must be overridden by the derived
+          class for each test statistic. Failure to do this will raise
+          ``TypeError``.
+
+        Args:
+          observed (:class:`numpy.ndarray`): 1D array containing the
+            observed data points.
+          expected (:class:`numpy.ndarray`): 1D array containing the
+            expected values predicted by the model.
+
+        Returns:
+          float: Computed value of test statistic.
+        """
+        return None
+
+
+class ChiSquared(TestStatistic):
+    """ Test statistic class for calculating the chi-squared test
+    satatistic.
+
+    Args:
+      spectra_par (:class:`echidna.core.spectra.SpectraParameter): The
+        parameter class for the projection dimension e.g.
+        ``energy_mc``.
+      form (string, optional): Form of chi-squared to use in calculation
+        of the test statistic. Default is "poisson_likelihood"
+      single_bin (bool, optional): If ``True``, the test statistic is
+        calculated for a single bin, i.e. just using the sum of
+        ``observed`` and ``expected``.
+
+    .. note:: All forms of chi-squared are as defined in::
+
+        * REF: S. Baker & R. D. Cousins, Nucl. Inst. and Meth. in Phys.
+          Res. 211, 437-442 (1984)
+
+      Available forms of chi-squared include::
+
+        * ``poisson_likelihood`` (default)
+        * ``pearson``
+        * ``neyman``
+
+    Raises:
+      TypeError: If class does not override abstract methods of base
+        class.
+    """
+    def __init__(self, spectra_par,
+                 form="poisson_likelihood", single_bin=False):
+        super(ChiSquared, self).__init__("chi_squared", spectra_par)
+        self._form = form
+        self._single_bin
+
+    def compute_statistic(self, observed, expected):
+        """ Compute the value of the test statistic.
+
+        Args:
+          observed (:class:`numpy.ndarray`): 1D array containing the
+            observed data points.
+          expected (:class:`numpy.ndarray`): 1D array containing the
+            expected values predicted by the model.
```

Comment:
Either that or I can just ravel the arrays as I pass them to you? Guess it makes no difference